FIPS/fip-draft_foundational_fvm.md

```diff
@@ -0,0 +1,1010 @@
+Rust to accompany this FIP. It is used by the canonical built-in actors.
+
+We find that Rust is a reasonable choice to writing native actors due to the
+lack of a runtime, the ability to elide the stdlib through `no_std`, and its
+mature WASM support.
+
+Exploration of other languages is something we encourage the community to
+pursue. Concretely, we believe that AssemblyScript, Swift, Kotlin, and TinyGo
+SDK are worth exploring.
+
+##### Foreign actors
+
+Foreign actors are smart contracts or programs targeting non-FVM runtimes,
+running inside the Filecoin network through emulation.
+
+The ability to run foreign actors is vital for cross-chain interoperability, and
+to enable other chains to access Filecoin storage capabilities more seamlessly.
+
+The first kind of foreign actor we aim to support is **Ethereum smart
+contract**, through the deployment of low-level EVM bytecode. This allows for
+straightforward and relatively risk-free porting of existing battle-tested
+Ethereum smart contracts to the Filecoin network.
+
+Every foreign runtime is in turn an actor type, identified by a `CodeCID`.
+Foreign runtimes are installed just like any other actor type. Foreign actors
+instances are constructed through the `InitActor`, supplying the source
+deployable as a constructor.
+
+In the case of the EVM foreign runtime, we are looking to standardise on either
+one of these interpreters:
+
+1. [SputnikVM], or
+2. [blueallow/revm] (a fork of the above)
+
+Every foreign runtime will need to specify bindings to map runtime-specific
+aspects (e.g. addressing, crypto primitives, gas accounting, etc.) to Filecoin
+counterparts. A preliminary [EVM <> FVM binding specification] exists.
+
+Future foreign runtime candidates include Agoric SES, Solana eBPF, and others.
+In all cases, compatibility should target the lowest level executable output in
+its source form, rather than dealing with high-level languages using
+alternative/custom toolchains. This reduces risk surface by increasing execution
+fidelity/parity, and makes it tooling reuse possible.
+
+#### Actor bytecode representation
+
+WASM bytecode payloads will be represented as an IPLD DAG, and will be stored in
+the state blockstore, owned by the node implementation.
+
+**The concrete DAG layout, multihash function, and CID codec are yet to be
+defined.**
+
+Code will be linked to from the state tree through the existing `code` CID field
+in the `ActorState` object:
+
+```rust
+pub struct ActorState {
+    /// Link to code for the actor.
+    pub code: Cid,
+    /// Link to the state of the actor.
+    pub state: Cid,
+    /// Current sequential nonce of the actor.
+    pub nonce: u64,
+    /// Filecoin tokens available to the actor.
+    pub balance: TokenAmount,
+}
+```
+
+#### Actor deployment
+
+The current `InitActor` (`f00`) will be extended with a (`InstallActor`) method
+taking two input parameters:
+
+1. The actor's WASM bytecode as an input parameter.
+2. The invocation container version.
```

Comment:
IMO, the wasm module should just specify this. We can add custom fields to the wasm metadata. Or, we could just version the runtime methods themselves.

Reply:
Do you mean as an exported global? That would work as long as there are no cases where we need to know the version _before_ compiling the module. (We _could_ locate the designated export by scanning the bytecode if we need to do that, but that's probably wasteful).